File: raft/grpc_server.go
package raft

import (
	"context"

	"github.com/eraft-io/eraft/raftpb"
)

type RaftgRPCServer struct {
	raftpb.UnimplementedRaftServiceServer
	rf *Raft
}

func NewRaftgRPCServer(rf *Raft) *RaftgRPCServer {
	return &RaftgRPCServer{rf: rf}
}

func (s *RaftgRPCServer) RequestVote(ctx context.Context, req *raftpb.RequestVoteRequest) (*raftpb.RequestVoteResponse, error) {
	args := &RequestVoteRequest{
		Term:         int(req.Term),
		CandidateId:  int(req.CandidateId),
		LastLogIndex: int(req.LastLogIndex),
		LastLogTerm:  int(req.LastLogTerm),
	}
	reply := &RequestVoteResponse{}
	s.rf.RequestVote(args, reply)
	return &raftpb.RequestVoteResponse{
		Term:        int64(reply.Term),
		VoteGranted: reply.VoteGranted,
	}, nil
}

func (s *RaftgRPCServer) AppendEntries(ctx context.Context, req *raftpb.AppendEntriesRequest) (*raftpb.AppendEntriesResponse, error) {
	entries := make([]Entry, len(req.Entries))
	for i, e := range req.Entries {
		entries[i] = Entry{
			Index:   int(e.Index),
			Term:    int(e.Term),
			Command: e.Command,
		}
	}
	args := &AppendEntriesRequest{
		Term:         int(req.Term),
		LeaderId:     int(req.LeaderId),
		PrevLogIndex: int(req.PrevLogIndex),
		PrevLogTerm:  int(req.PrevLogTerm),
		LeaderCommit: int(req.LeaderCommit),
		Entries:      entries,
	}
	reply := &AppendEntriesResponse{}
	s.rf.AppendEntries(args, reply)
	return &raftpb.AppendEntriesResponse{
		Term:          int64(reply.Term),
		Success:       reply.Success,
		ConflictIndex: int64(reply.ConflictIndex),
		ConflictTerm:  int64(reply.ConflictTerm),
	}, nil
}

func (s *RaftgRPCServer) InstallSnapshot(ctx context.Context, req *raftpb.InstallSnapshotRequest) (*raftpb.InstallSnapshotResponse, error) {
	args := &InstallSnapshotRequest{
		Term:              int(req.Term),
		LeaderId:          int(req.LeaderId),
		LastIncludedIndex: int(req.LastIncludedIndex),
		LastIncludedTerm:  int(req.LastIncludedTerm),
		Data:              req.Data,
	}
	reply := &InstallSnapshotResponse{}
	s.rf.InstallSnapshot(args, reply)
	return &raftpb.InstallSnapshotResponse{
		Term: int64(reply.Term),
	}, nil
}

File: raft_cpp/src/grpc_server.cpp
// grpc_server.cpp — gRPC service implementation
// Corresponds to Go: grpc_server.go

#include "raft/grpc_server.h"

namespace raft {

grpc::Status RaftgRPCServiceImpl::RequestVote(
    grpc::ServerContext* /*context*/,
    const raftpb::RequestVoteRequest* req,
    raftpb::RequestVoteResponse* resp)
{
    RequestVoteRequest args;
    args.term          = static_cast<int>(req->term());
    args.candidate_id  = static_cast<int>(req->candidate_id());
    args.last_log_index = static_cast<int>(req->last_log_index());
    args.last_log_term  = static_cast<int>(req->last_log_term());

    RequestVoteResponse reply;
    rf_->HandleRequestVote(args, reply);

    resp->set_term(reply.term);
    resp->set_vote_granted(reply.vote_granted);
    return grpc::Status::OK;
}

grpc::Status RaftgRPCServiceImpl::AppendEntries(
    grpc::ServerContext* /*context*/,
    const raftpb::AppendEntriesRequest* req,
    raftpb::AppendEntriesResponse* resp)
{
    AppendEntriesRequest args;
    args.term          = static_cast<int>(req->term());
    args.leader_id     = static_cast<int>(req->leader_id());
    args.prev_log_index = static_cast<int>(req->prev_log_index());
    args.prev_log_term  = static_cast<int>(req->prev_log_term());
    args.leader_commit  = static_cast<int>(req->leader_commit());

    for (const auto& pe : req->entries()) {
        Entry e;
        e.index = static_cast<int>(pe.index());
        e.term  = static_cast<int>(pe.term());
        if (!pe.command().empty()) {
            e.command.assign(pe.command().begin(), pe.command().end());
        }
        args.entries.push_back(std::move(e));
    }

    AppendEntriesResponse reply;
    rf_->HandleAppendEntries(args, reply);

    resp->set_term(reply.term);
    resp->set_success(reply.success);
    resp->set_conflict_index(reply.conflict_index);
    resp->set_conflict_term(reply.conflict_term);
    return grpc::Status::OK;
}

grpc::Status RaftgRPCServiceImpl::InstallSnapshot(
    grpc::ServerContext* /*context*/,
    const raftpb::InstallSnapshotRequest* req,
    raftpb::InstallSnapshotResponse* resp)
{
    InstallSnapshotRequest args;
    args.term               = static_cast<int>(req->term());
    args.leader_id          = static_cast<int>(req->leader_id());
    args.last_included_index = static_cast<int>(req->last_included_index());
    args.last_included_term  = static_cast<int>(req->last_included_term());
    if (!req->data().empty()) {
        args.data.assign(req->data().begin(), req->data().end());
    }

    InstallSnapshotResponse reply;
    rf_->HandleInstallSnapshot(args, reply);

    resp->set_term(reply.term);
    return grpc::Status::OK;
}

} // namespace raft
